### PR TITLE
[ownership] Treat open_existential_box as an interior pointer producting instruction.

### PR DESCRIPTION
**Pull request by gottesmm**
This means that it can only have a guaranteed object as an operandand that we
validate that all uses of the result address of open_existential_box occur only
within the lifetime of said object's borrow scope.

----

I am most interested in review about the SILGen part.

### CODE REVIEW

File: lib/SILGen/SILGenPoly.cpp [hunk 1]
```diff
@@ -4141,6 +4141,7 @@ void SILGenFunction::emitProtocolWitness(AbstractionPattern reqtOrigTy,
     break;
   }
 
+  formalEvalScope.pop();
```
Comment (gottesmm) [inline lib/SILGen/SILGenPoly.cpp]: We weren't ever popping the formal eval scope... Was causing my conversion to use formal access cleanups in certain places to trigger the assert about mixing formal evaluation/non-formal evaluation scopes in appropriately.

File: lib/SILGen/SILGenLValue.cpp [hunk 2]
```diff
@@ -873,9 +873,14 @@ namespace {
           auto &TL = SGF.getTypeLowering(base.getType());
           error = SGF.emitLoad(loc, base.getValue(), TL,
                                SGFContext(), IsNotTake);
+          // Error comes back to us with a +1 cleanup that is not a formal
+          // access cleanup. We need it to be that so that the load nests
+          // properly with other lvalue scoped things like the borrow below.
+          error = SGF.emitFormalAccessManagedRValueWithCleanup(
```
Comment (gottesmm) [inline lib/SILGen/SILGenLValue.cpp]: As I was looking at this code, I noticed that it is in SILGenLValue where everything /should/ have a formal access duration. I noticed that emitLoad here was always returning a normal load. So, I put in some code here to ensure that we forward that non-formal access release and create a formal access release. That is why in some of the test updates, a destroy_value is hoisted earlier.

**Review by gottesmm — COMMENTED** [review]
Two quick notes on the SILGen stuff.

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
@swift-ci test source compatibility

**Comment by gottesmm** [comment]
@swift-ci benchmark

**Comment by gottesmm** [comment]
@swift-ci clean test os x platform

**Comment by gottesmm** [comment]
@swift-ci test os x platform

**Review by jckarter — APPROVED** [review]
Makes sense, it should be isomorphic to `ref_element_addr` for most purposes.

**Pull request merged**